### PR TITLE
Update the logging properties to opt-out of the prefix events #844

### PR DESCRIPTION
**Pull request by mickeyz07**
Determines whether ESAPI should log the prefix of [EVENT_TYPE - APPLICATION NAME].
If all Logger.* settings in the ESAPI.properties file are false (including Logger.LogPrefix), then the log output would be the same as if no ESAPI was used.
That can be important, especially for clients that are using cloud providers. The charges for extra logging can quickly add up.
So, the user can sanitise untrusted data used to construct log entries by using a safe logging mechanism in the OWASP ESAPI Logger without any extra logging information.
However, only by changing a couple of settings in the ESAPI.properties file, a user can have a very detailed log if required.

### CODE REVIEW

File: configuration/esapi/ESAPI.properties
Comment (jeremiahjstacey) [inline configuration/esapi/ESAPI.properties:412]: The boolean logic / variable name is a little confusing.

I think the logic will be easier to manage if we phrase this as a Positive statement:
```
# Determines whether ESAPI should log the prefix of [EVENT_TYPE - APPLICATION NAME].
Logger.LogPrefix=true  
```

I'd prefer to remove the additional comment from line 411, as that may lead to confusion around the "other log entries" if (and when) the properties are expanded.

Changing the naming to a positive statement will also apply throughout the naming of the implementation, and the  boolean logic checks.

File: src/main/java/org/owasp/esapi/logging/appender/ServerInfoSupplier.java
Comment (kwwall) [inline src/main/java/org/owasp/esapi/logging/appender/ServerInfoSupplier.java:63]: @jeremiahjstacey and @xeno6696 - Seeing this common boilerplate reminds me to let's make a note to use Lombok annotations in ESAPI 3.x. I'd do it now, but I really don't want to add another dependency at this point until we can break things up into smaller components and reduce the total dependencies where you are just using small pieces.

File: src/test/java/org/owasp/esapi/logging/appender/EventTypeLogSupplierIgnoreEventTypeTest.java
Comment (kwwall) [inline src/test/java/org/owasp/esapi/logging/appender/EventTypeLogSupplierIgnoreEventTypeTest.java:37]: Suggestion: Add blank line here for improved readability.

File: src/test/resources/esapi/ESAPI.properties
Comment (kwwall) [inline src/test/resources/esapi/ESAPI.properties:444]: Change this property name in accordance with @jeremiahjstacey's suggestion in c**onfiguration/esapi/ESAPI.properties.**

**Review by kwwall — COMMENTED** [review]
So far, no objections, but I need to think about the tests a bit more. In the meantime, @mickeyz07, perhaps you can respond to my comments.

File: src/main/java/org/owasp/esapi/logging/appender/EventTypeLogSupplier.java
Comment (jeremiahjstacey) [inline src/main/java/org/owasp/esapi/logging/appender/EventTypeLogSupplier.java:51]: Preferred to have single returns from methods. 
Ternary recommended (please note that my sample is applying the updated property name):
`return logEventType ? eventType.toString() ? "";`

File: src/main/java/org/owasp/esapi/logging/appender/LogPrefixAppender.java
Comment (jeremiahjstacey) [inline src/main/java/org/owasp/esapi/logging/appender/LogPrefixAppender.java:56]: I'd be in favor of removing this constructor entirely. That can have impacts if anyone has actually extended the class, so I suspect that we would not want to do that.

Assuming that we cannot delete this code, I'd like to request two more adjustments here:
1.  Update the Javadoc to specify the logPrefix value is configured to true
2. Chain the constructors - have this one call the other
```
public LogPrefixAppender(boolean logUserInfo, boolean logClientInfo, boolean logServerIp, boolean logApplicationName, String appName) {
  this (logUserInfo, logClientInfo, logServerIp, logApplicationName, appName, true);
}
```

File: src/main/java/org/owasp/esapi/logging/appender/ServerInfoSupplier.java
Comment (jeremiahjstacey) [inline src/main/java/org/owasp/esapi/logging/appender/ServerInfoSupplier.java:68]: this feels incorrect.
If the applicationname is null then we want to log it?

File: src/test/java/org/owasp/esapi/logging/appender/LogPrefixAppenderTest.java
Comment (jeremiahjstacey) [inline src/test/java/org/owasp/esapi/logging/appender/LogPrefixAppenderTest.java:37]: clean up commented line

File: src/main/java/org/owasp/esapi/logging/appender/ServerInfoSupplier.java
Comment (jeremiahjstacey) [reply src/main/java/org/owasp/esapi/logging/appender/ServerInfoSupplier.java:68]: Looking closer at this,  I'm not sure the inner if/elsif block adds value?

//If not null and not empty -> Log the app name
// else if null -> Log the app name

I think it makes sense to remove the  applicationName validation and  log the applicationName if it's configured to log.

```
if (this.logAppName) {
  appInfo.append("/").append(this.applicationName);
}
```

If the log does not appear correctly, that is a different configuration that should be made in the environment.

Comment (kwwall) [reply src/main/java/org/owasp/esapi/logging/appender/ServerInfoSupplier.java:68]: If `applicationName` is `null` or the empty string, perhaps we should log it as "Unknown"?

Comment (jeremiahjstacey) [reply src/main/java/org/owasp/esapi/logging/appender/ServerInfoSupplier.java:68]: If that were the case, then we'd want to change the default value returned from the SecurityConfiguration so that an unset ApplicationName would be consistent throughout the instance.
I think that's out of scope for this work, and I think we should log the value that was pulled from the config (even if that is null). 
Note that the ApplicationName is pulled from the SecurityConfiguration in each of the LogFactory implementations:
` String appName = ESAPI.securityConfiguration().getStringProp(APPLICATION_NAME);`

**Comment by jeremiahjstacey** [comment]
@kwwall - This all looks good to me.  if your concerns have been addressed please merge or leave a comment and I will follow up and get it merged this week

File: src/main/java/org/owasp/esapi/logging/appender/ServerInfoSupplier.java
Comment (kwwall) [reply src/main/java/org/owasp/esapi/logging/appender/ServerInfoSupplier.java]: @jeremiahjstacey - Fair enough. But maybe something we should keep in mind for the future.

**Comment by kwwall** [comment]
> @kwwall - This all looks good to me. if your concerns have been addressed please merge or leave a comment and I will follow up and get it merged this week

@jeremiahjstacey - I will take care of merging this, but probably won't get to it by today though, as I had hoped. But I want to run a few dynamic tests first to make sure the defaults are all set so as to be backward compatible. I think they are, but it's a lot easier to try it than try to judge based on the code. I hope to have it merged by this weekend unless I find something that needs fixed

**Comment by kwwall** [comment]
@mickeyz07 - Sorry I didn't notice this before. Jeremiah and I were focused on the code review itself. However, I just noticed that you have not signed your commits with a GPG / PGP key. For commits that will be merged to the 'develop' or 'main' branches, we require those commits to be signed.

From our **CONTRIBUTING-TO-ESAPI.txt** file:
```text
A Special Note Regarding Making Commits for PRs
    Shortly after the 2.5.1.0 ESAPI release in late November 2022, the ESAPI
    team decided to lock down the 'develop' amd 'main' branches. Merges from
    PRs are done to the 'develop' branch. That means that if you intend to
    contribute to ESAPI, you must be signing your commits. Please see the
    GitHub instructions at
        https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits
    for details.
```

There are detailed instructions on how to set this up in our [ESAPI Release Steps document.](https://github.com/ESAPI/esapi-java-legacy/blob/develop/documentation/ESAPI-release-steps.pdf) Unfortunately, I seem to recall this before. GitHub doesn't warn us about this until after we have 2 people from the team approve the PR. I think it would then probably then force you to go back and set up signing and you'd have to made at least one commit to all that you've modified and then we'd have to get 2 of us to approve it again since there would be a commit after the approval.

So, if you could, could you take care of that. I'm not sure if it would fix it though, because I think ti may require ALL commits to be signed, not just the last one in a PR. (@xeno6696 or @jeremiahjstacey - do either of you know?)

I'm guessing it may not have flagged this because you forked it and created it on your 'issue-844' branch (which, was as instructed). I think there is an option where I can set up our repo to require signed commits on ALL branches, but I didn't do that because I figured it my break things like Dependabot and Snyk generated PRs.

Anyhow, can you see if you can somehow fix this? Thanks.

**Comment by xeno6696** [comment]
There's no reason we couldn't have a "signature" commit, just a simple merge with a text change detailing it as a signature.  Just to get it into history and I think that ought to be sufficient.  Going back and resigning old commits is more trouble than its worth.
________________________________

**Comment by kwwall** [comment]
@xeno6696 - That's sort of what I was thinking of having him do, but I wasn't sure if that would pass the muster of the GitHub signature verification. I thought it required every single commit to be signed if it was to end up on the 'develop' or 'main' branch. That said, maybe it would work if we did a squash-merge? I don't know.

**Comment by kwwall** [comment]
So, @mickeyz07 found a severe **must fix** issue in your code. I made this simple change to simulate users NOT updating their ESAPI.properties file (which I can guarantee you that many will do):
```bash
$ git diff src/test/resources/esapi/ESAPI.properties
diff --git a/src/test/resources/esapi/ESAPI.properties b/src/test/resources/esapi/ESAPI.properties
index 8ffc61f6..a84328ab 100644
--- a/src/test/resources/esapi/ESAPI.properties
+++ b/src/test/resources/esapi/ESAPI.properties
@@ -441,7 +441,7 @@ Logger.ClientInfo=true
 
 # Determines whether ESAPI should log the prefix of [EVENT_TYPE - APPLICATION NAME].
 # If all above Logger entries are set to false, as well as LogPrefix, then the output would be the same as if no ESAPI was used
-Logger.LogPrefix=true
+# Logger.LogPrefix=true
 
 #===========================================================================
 # ESAPI Intrusion Detection
```
And when I re-ran 'mvn test' I got this:
```text
[ERROR] Tests run: 4256, Failures: 7, Errors: 2434, Skipped: 0
```
A similar error happens if one sets the value of `Logger.LogPrefix` to anything other than `true` or `false`. They are all throwing ConfigurationExceptions some of which end up as `ExceptionInInitializerError`. That's why I said it was important that regardless of the value, I want to treat the absence of this property value or any invalid property value as it was in 2.5.4.0. So, it they put in:
```bash
Logger.LogPrefix=off
```
or even something absurd as:
```bash
Logger.LogPrefix=scoopy_do
```
you have to treat is as though they had:
```bash
Logger.LogPrefix=true
```
It's okay if you want to output an error to stdout using `logSpecial` or whatever it is that we normally use for that, but we can't be throwing exceptions that will kill their application just because they haven't followed instructions. Because many won't.

Make sense?

**Review by kwwall — CHANGES_REQUESTED** [review]
Changes discussed in https://github.com/ESAPI/esapi-java-legacy/pull/845#issuecomment-2211793935 are required before this will be approved.

**Comment by mickeyz07** [comment]
@kwwall @jeremiahjstacey I have made changes and signed the commit with GPG key.
Currently if any of the Logger.* properties is missing or it has value different then true, false, yes, no the tests are failing. 
So to reiterate - the problem is not only with the Logger.LogPrefix.
As you could see in the code I have made provision for all Logger.* properties. 
Only Logger.LogEncodingRequired=false.
All other Boolean Logger.* properties have the default value of true.
Not sure if you require to extend that behavior to other properties? 
So if the value for Boolean properties is not valid, instead of throwing an exception, now reverts the value for the property to the default value?
In that case we could have another function with additional default parameter:
public Boolean getBooleanProp(String propertyName, boolean value)
That  would require additional testing.
So for this PR, the changes that I made are taking care of the Logger.* properties.
All tests that Kevin "broke" with commenting out Logger.* are now working correctly.

File: src/main/java/org/owasp/esapi/reference/DefaultSecurityConfiguration.java
Comment (kwwall) [inline src/main/java/org/owasp/esapi/reference/DefaultSecurityConfiguration.java:1447]: @mickeyz07 -- IMO, this is a major implementation kludge to me. Generally, I was expecting in whatever static initializer that checks the value for the `Logger.LogPrefix` property, that any error / exception that would be thrown that you would just treat it as 'true' and then log something via `logSpecial`. If for some reason that approach turns out to be impossible, then the next best alternative would be to create a special dedicated getLogPrefix method. Another preferred alternative (something I've been meaning to do for a while now) is to create a new `Boolean getBooleanProp(String propName, boolean default)` and use that, and just pass in 'true' for the default, so if the specified property name is not a boolean, it defaults to the specified default value (which should be true for `Logger.LogPrefix.` I am also very confused why this is even written the way that it it; it appears that you could omit the entire:
```java
                if (propertyName.startsWith("Logger.")) {
```
and just went with the specific property name check that follows. (And why _this_ specific check? I would have expected `Logger.LogPrefix` here, since that's the new property. So why `Logger.LogEncodingRequired`? I'm confused.)

@jeremiahjstacey - I'd like your opinion on this, but I don't like this **_at all._** Can your think of any alternatives.

**Review by kwwall — COMMENTED** [review]
Not going to approve this unless Matt or Jeremiah find this kludge acceptable.

File: src/main/java/org/owasp/esapi/reference/DefaultSecurityConfiguration.java
Comment (jeremiahjstacey) [reply src/main/java/org/owasp/esapi/reference/DefaultSecurityConfiguration.java:1447]: I think I agree @kwwall  - Updates in this file will lead to additional scope that I think we'll be better off not trying to tackle in this effort.

@mickeyz07 - My understanding of what the desire here is to try to solve the problem at the lowest level possible.  I think I agree with the intent, but I think you can agree that it's not the cleanest or most maintainable due to current restrictions in the library's interfaces.
To facilitate a faster integration of this feedback, I would like to propose the following:
1.  Revert all changes to DefaultSecurityConfiguration.java
2. Update JavaLogFactory and SLF4JLogFactory implementations to default to TRUE, attempt to read from properties, and useLogSpecial on failure

```
boolean logPrefix = true;
try {
  logPrefix = ESAPI.securityConfiguration().getBooleanProp(LOG_PREFIX);
} catch (ConfigurationException ex) {
ESAPI.securityConfiguration().logSpecial("Failed to read Log Prefix configuration. Defaulting to enabled. ", ex);
}
```

It's not the best possible solution, but I think that's the best we can do in the current library implementation.

Comment (kwwall) [reply src/main/java/org/owasp/esapi/reference/DefaultSecurityConfiguration.java:1447]: Okay, my thoughts, exactly, and what you described was my preferred approach at resolving this.

@mickeyz07 - Please make the change that @jeremiahjstacey detailed above. Thanks.

**Review by kwwall — CHANGES_REQUESTED** [review]
Changes requested. See previous comments by Jeremiah and myself. It should be pretty straightforward since he pretty much laid out the code changes.

**Comment by mickeyz07** [comment]
@kwwall @jeremiahjstacey I have added the new method getBooleanProp(String propertyName, boolean value)
I have used that method only for Logger.LogPrefix, e.g.
        boolean logServerIp = ESAPI.securityConfiguration().getBooleanProp(LOG_SERVER_IP);
        boolean logPrefix = ESAPI.securityConfiguration().getBooleanProp(LOG_PREFIX, true);
All the tests run successfully.

**Comment by kwwall** [comment]
@mickeyz07 We are getting close, so please don't give up. But there are 2 things that I see missing right out of the gate:

1. You need to add the new `getBooleanProp` to src/test/java/org/owasp/esapi/SecurityConfigurationWrapper.java
2. You need to add some JUnit tests for the new method. Probably put those tests in the following files where other tests for the original `getBooleanProp` method resides:
    * src/test/java/org/owasp/esapi/configuration/EsapiPropertyManagerTest.java
    * src/test/java/org/owasp/esapi/configuration/StandardEsapiPropertyLoaderTest.java
    * src/test/java/org/owasp/esapi/configuration/XmlEsapiPropertyLoaderTest.java

**Comment by mickeyz07** [comment]
@kwwall I have added the required tests for all 3 configuration test files you noted. Please check.

File: src/main/java/org/owasp/esapi/configuration/EsapiPropertyLoader.java
Comment (kwwall) [inline src/main/java/org/owasp/esapi/configuration/EsapiPropertyLoader.java:38]: From the Javadoc: for this interface: "or has incorrect type". 

Well, that is the $64k question, isn't it? What is the _INTUITIVE_ behavior if one calls something like:
```java
    Boolean propValue = getBoolean("SomePropName", true)
```
but somewhere, say in **ESAPI.properties** or elsewhere, we have defined
```bash
SomePropName=This is not a boolean property
```
Then what is the expected / intuitive behavior for that call? I think that I'd expect a `ConfigurationException` to be thrown rather than having `propValue` silently being set to `true`. Or maybe I shouldn't say that is the _intuitive_ behavior, but rather the _**safe**_ behavior. I think because copy/paste errors abound and it would be easy to make a typo with the property name, that's the only way that we're going to help people catch these bizarre accidental edge cases. Otherwise, we're going to cause them to curse at us because they'll have to figure up a debugger to figure out what the hell is going on in cases like this.

@xeno6696 and @jeremiahjstacey  - I'd like your opinion here. I know this requires yet another code change, but if we do a release and document it this way this (likely bug, IMO) will become a feature because changing it later on would break backward compatibility. If we were requiring JDK 9 or later we could just not export this method and keep it hidden, but since Java 8 is the minimal version, once we add this, it becomes available to the world.

File: src/main/java/org/owasp/esapi/configuration/EsapiPropertyManager.java
Comment (kwwall) [inline src/main/java/org/owasp/esapi/configuration/EsapiPropertyManager.java:85]: I think the way this is currently implemented, I don't think it will ever reach the final return on line 87 here. Maybe I'm wrong about that. Do we have a test for that case? Of course, if we change it as per my previous comment, we may still need it. Or may need to add a try / catch here and just rethrown any `ConfigurationException`.

File: src/main/java/org/owasp/esapi/configuration/StandardEsapiPropertyLoader.java
Comment (kwwall) [inline src/main/java/org/owasp/esapi/configuration/StandardEsapiPropertyLoader.java:88]: Depending on if we decide to change the behavior here in accordance to my first comment, we may wish to change line 88 to something along the following lines:
```java
    if ( property != null ) {
        throw new ConfigurationException("Property name '" + propertyName + "' has non-Boolean value of '" + property + "'.");
    }
```

File: src/main/java/org/owasp/esapi/configuration/XmlEsapiPropertyLoader.java
Comment (kwwall) [inline src/main/java/org/owasp/esapi/configuration/XmlEsapiPropertyLoader.java:104]: Ditto here. Similar to comment on **StandardEsapiPropertyLoader.java**.

File: src/main/java/org/owasp/esapi/reference/DefaultSecurityConfiguration.java
Comment (kwwall) [inline src/main/java/org/owasp/esapi/reference/DefaultSecurityConfiguration.java:1481]: Ditto comment in **EsapiPropertyManager.java** about throwing `ConfigurationException` here.

File: src/test/java/org/owasp/esapi/configuration/EsapiPropertyManagerTest.java
Comment (kwwall) [inline src/test/java/org/owasp/esapi/configuration/EsapiPropertyManagerTest.java:384]: If we decide to revise this behavior to throw a `ConfigurationException` if `getBooleanProperty(String, Boolean)` is called and a non-Boolean property if found, we will need a new test here.

Ditto in **StandardEsapiPropertyLoaderTest.java** and **XmlEsapiProertperLoaderTest.java**.

**Review by kwwall — COMMENTED** [review]
@mickeyz07 - I'm holding off on requesting changes until I get @xeno6696 and @jeremiahjstacey feedback about how getBooleanProp(String, Boolean) should treat a property that it encounters that has a non-Boolean value. I think we need to resolve that first. If we decide to do that, I will give you the option to finish it up, but if you don't, I'll probably approve this and then do those final changes myself in a separate PR. But getting this behavior of this new method is important and we'll likely only get one chance at getting it right.

@xeno6696 and @jeremiahjstacey -- waiting to hear back from you on this.

File: src/main/java/org/owasp/esapi/configuration/EsapiPropertyLoader.java
Comment (jeremiahjstacey) [reply src/main/java/org/owasp/esapi/configuration/EsapiPropertyLoader.java:38]: Overall, I think this is more scope than we should put time into to modify the opt-out request from the original issue.  I appreciate the initiative; however, we're adding this case to expand for something that could have been solved in two copies of an 6 line block.

This is overengineering in this scope.  We're blocking value, and a new possible release for this API change.

My suggestion is to take it out.  Use the duplicate snippet from my previous comment in the 2 factories and revert the changes in the DefaultSecurityConfiguration.  Create another task for the default API that we can address (reuse this code if you want), and complete it in a non-blocking scope.  We don't need this to fix the original problem, and as you can see the discussion of edge cases and caveats will continue to grow.  Let's get the opt-out behavior in in a first form, then improve upon it in another task to expand the SecurityConfiguration API to allow defaults to be specified.

Comment (kwwall) [reply src/main/java/org/owasp/esapi/configuration/EsapiPropertyLoader.java:38]: I concur. If in the future, you want to do a new PR to get a new interface for
```java
public Boolean getBooleanProp(String propertyName, Boolean defaultValue) throws ConfigurationException
```
I am fine with that, but as my review of this as surfaced, this is a little more nuanced than intuition alone might have us believe. If we are going to add that a future time, I think we need to start out with a GitHub issue and discuss the specifications there first, before waiting for an implementation to discuss them. So, to proceed with this PR, please rip that part of the code out and follow Jeremiah's suggestion.

File: src/main/java/org/owasp/esapi/logging/java/JavaLogFactory.java
Comment (kwwall) [inline src/main/java/org/owasp/esapi/logging/java/JavaLogFactory.java:88]: Please change the output to something that explicitly mentions the property name on the output sent to stderr. E.g., something like this:
```java
 System.out.println("ESAPI: Failed to read Log Prefix configuration, " + LOG_PREFIX + ". Defaulting to enabled" +
```
That will make it easier for those getting this failure to know exactly what property to look for if they wish to change it in their **ESAPI.properties** file.

File: src/main/java/org/owasp/esapi/logging/slf4j/Slf4JLogFactory.java
Comment (kwwall) [inline src/main/java/org/owasp/esapi/logging/slf4j/Slf4JLogFactory.java:79]: Same as previous comment.

**Review by kwwall — CHANGES_REQUESTED** [review]
Two lines where I'd like to see some very minor tweaks, otherwise LGTM.

**Review by kwwall — APPROVED** [review]
LGTM.

**Review by jeremiahjstacey — APPROVED** [review]
Looks good to me as well

**Comment by kwwall** [comment]
I'm going to go ahead and bypass the 'branch protection' and do the 'squash and merge' because @mickeyz07 signed most of his commits and we reviewed all the code, so someone usurped his GitHub username and snuck something malicious past us that's one us.

**Pull request merged**
